hiero-observability/hiero-metrics-core/src/main/java/org/hiero/metrics/api/core/StatefulMetric.java

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+package org.hiero.metrics.api.core;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Base interface for a metric that can have multiple data points holding values,
+ * each associated with a unique set of dynamic label values.
+ * <p>
+ * Implementation is responsible for creating a new data point for each unique set of label values.
+ * <p>
+ * Clients should pay attention the dynamic label values cardinality, as high cardinality can lead to
+ * higher costs for metrics backends. <b>Do not use</b> labels with values having unbounded cardinality,
+ * such as IDs or timestamps.
+ * <p>
+ * Data points are created lazily, when requested via {@link #getOrCreateLabeled(String...)},
+ * {@link #getOrCreateLabeled(Object, String...)} or {@link #getOrCreateNotLabeled()} for observation,
+ * so no data point will be exported until at least one observation is made using these methods.
+ * If initial value of data point has to be exported even before observation,
+ * {@code getOrCreate} method can be called to instantiate datapoint.
+ *
+ * @param <I> the type of the initializer used to create new data points per label set
```

Comment:
Could you provide some more details on the initializer, please? Is it to create new data points for this metric? If so, could it be just `Supplier<D>` rather than a totally different type `I`?